rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -528,100 +606,172 @@ def slice_state_history(
+        )
+        return self.training_states_labels_and_entities(
+            trackers, domain, omit_unset_slots=omit_unset_slots
+        )
+
+    def training_states_labels_and_entities(
+        self,
+        trackers: List[DialogueStateTracker],
+        domain: Domain,
+        omit_unset_slots: bool = False,
+    ) -> Tuple[List[List[State]], List[List[Text]], List[List[Dict[Text, Any]]]]:
+        """Transforms list of trackers to lists of states, labels and entity data.
+
+        Args:
+            trackers: The trackers to transform
+            domain: The domain
+            omit_unset_slots: If `True` do not include the initial values of slots.
+
+        Returns:
+            A tuple of list of states, list of labels and list of entity data.
+        """
+
+        self._setup_example_iterator()
 
-        # from multiple states that create equal featurizations
-        # we only need to keep one.
-        hashed_examples = set()
+        example_states = []
+        example_labels = []
+        example_entities = []
 
         logger.debug(
-            "Creating states and action examples from "
+            "Creating states and {} label examples from "
             "collected trackers (by {}({}))..."
-            "".format(type(self).__name__, type(self.state_featurizer).__name__)
+            "".format(
+                self.LABEL_NAME,
+                type(self).__name__,
+                type(self.state_featurizer).__name__,
+            )
         )
         pbar = tqdm(
             trackers,
             desc="Processed trackers",
             disable=rasa.shared.utils.io.is_logging_disabled(),
         )
         for tracker in pbar:
-            states = self._create_states(
-                tracker, domain, omit_unset_slots=omit_unset_slots
-            )
 
-            states_length_for_action = 0
-            entity_data = {}
-            for event in tracker.applied_events():
-                if isinstance(event, UserUttered):
-                    entity_data = self._entity_data(event)
+            for states, label, entities in self._example_iterator(
+                tracker, domain, omit_unset_slots=omit_unset_slots
+            ):
 
-                if not isinstance(event, ActionExecuted):
+                if self._check_example_cache(tracker, states, label):
                     continue
 
-                states_length_for_action += 1
+                example_states.append(states)
+                example_labels.append(label)
+                example_entities.append(entities)
+
+                pbar.set_postfix({f"# {self.LABEL_NAME}": f"{len(example_labels):d}"})
+
+        self._cleanup_example_iterator()
+        self._remove_user_text_if_intent(example_states)
+
+        logger.debug(f"Created {len(example_states)} {self.LABEL_NAME} examples.")
+
+        return example_states, example_labels, example_entities
+
+    def _setup_example_iterator(self) -> None:
```

Comment:
I am not a fan of these iterator setup and cleanup procedures. This reliance on side effects makes the code very path dependent and difficult to understand locally, debug, and reuse. It seems to me, it should be possible to put the whole filtering of duplicates into the `_example_iterator` method so that it doesn't need to rely on instance variables that have to be setup independently

Reply:
Will remove these functions with side effects and bring them inside `training_states_labels_and_entities`.